envoy/common/inline_map_registry.h

```diff
@@ -0,0 +1,513 @@
+#pragma once
+
+#include <cstddef>
+#include <cstdint>
+
+#include "source/common/common/assert.h"
+#include "source/common/common/macros.h"
+#include "source/common/common/utility.h"
+
+#include "absl/container/flat_hash_map.h"
+#include "absl/container/node_hash_set.h"
+
+namespace Envoy {
+
+class UntypedInlineHandle {
+public:
+  UntypedInlineHandle(uint64_t inline_scope_id, uint64_t inline_entry_id,
+                      absl::string_view inline_entry_key)
+      : inline_scope_id_(inline_scope_id), inline_entry_id_(inline_entry_id),
+        inline_entry_key_(inline_entry_key) {}
+
+  uint64_t inlineScopeId() const { return inline_scope_id_; }
+
+  uint64_t inlineEntryId() const { return inline_entry_id_; }
+
+  absl::string_view inlineEntryKey() const { return inline_entry_key_; }
+
+private:
+  uint64_t inline_scope_id_{};
+  uint64_t inline_entry_id_{};
+  absl::string_view inline_entry_key_;
+};
+
+/**
+ * Information about the registry. This is used to print out the registry debug information.
+ */
+struct InlineMapRegistryDebugInfo {
+  // Registry scope id for the registry, this is generated dynamically by the
+  // InlineMapRegistryManager.
+  uint64_t registry_scope_id;
+  // Registry scope name for the registry, typically short string from
+  // scope type.
+  absl::string_view registry_scope_name;
+  // All inline keys registered for the registry.
+  std::vector<absl::string_view> registry_inline_keys;
+};
+
+class InlineMapRegistryManager {
+public:
+  static const std::vector<InlineMapRegistryDebugInfo>& registryInfos() {
+    // Call finalize() to ensure that all finalizers are called.
+    finalize();
+    return mutableRegistryInfos();
+  }
+
+  /**
+   * Generate a scope id for the given scope and register a finalizer to be called when
+   * InlineMapRegistry::finalize() is called.
+   *
+   * NOTE: This is called by InlineMapRegistry::scopeId() and should never be called
+   * manually.
+   */
+  template <class Scope> static uint64_t generateScopeId();
+
+private:
+  using Finalizer = std::function<InlineMapRegistryDebugInfo()>;
+
+  static uint64_t nextScopeId() {
+    static uint64_t next_scope_id = 0;
```

Comment:
I would like to stop here and lobby strongly for this class not to lock down its usage model being static objects.

If a *caller* wants to put one in an evil singleton they can, but I think this class should not impose that storage model on callers.

In particular, next_scope_id looks like a thread issue in that context. I'd like for the registry object to have all the required data as member variables, and a caller can make a registry singleton or put one in a thread-local instance or whatever.

Classes can be thread-safe, thread-compatible, or thread-hostile. As written this is thread-hostile. https://faithlife.codes/blog/2008/03/degrees_of_thread_safety/ has some description of these terms. I would like this class to be thread-compatible. This is how STL works also -- std::map can be used in thread-safe code but have to do your own sync.

As an anti-pattern, I would call c_ares (Envoy's DNS library) thread hostile. You need to do init/destruct calls from a single thread, and the main reason is that c_ares has mutable statics.

Reply:
I think according the design, the `InlineMapRegistryManager` can only be used in the main thread to finalize and collect all registered registry‘s info.

But I can remove all scope id related code. The scope id is only used to support the untyped inline handle, but it's a optional enhancement and I can remove it first to make the PR more clean.